### PR TITLE
WebSocket server implementation

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR implements a basic WebSocket server. It's not 100% complete but PR was getting larger so opted to send it now.

Some important things still missing:

- Documentation for exported methods and types.
- Connection timeout handler (ping-pong).
- More tests.

The API is rather simple, some highlights:

- Server implements the `http.Handler` to be easily hooked to the existing HTTP API server without need for coupling.
- Higher level won't be dealing with connections directly but through message passing (using connection ids).
- Two channels of communication, one to send and one to receive messages.

### CODE REVIEW

File: service/ws/server_test.go [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+	s, addr, shutdown := setupServer(t, WithUpgradeCb(upgradeCb))
+	defer shutdown()
+	defer s.Close()
+
+	_, port, err := net.SplitHostPort(addr)
+	require.NoError(t, err)
+
+	u := url.URL{Scheme: "ws", Host: "localhost:" + port, Path: "/ws"}
+	c, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
```
Comment (cpoile) [inline service/ws/server_test.go:101]: cool!

File: service/ws/utils.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+const charset = "ybndrfg8ejkmcpqxot1uwisza345h769"
+
+var encoding = base32.NewEncoding(charset)
+
+// NewID is a globally unique identifier.  It is a [A-Z0-9] string 26
+// characters long.  It is a UUID version 4 Guid that is zbased32 encoded
+// with the padding stripped off.
+func newID() string {
```
Comment (cpoile) [inline service/ws/utils.go:20]: any reason we're not using `model.NewId()`?

**Review by cpoile — APPROVED** [review]
Awesome!

File: service/ws/utils.go (hunk 2 above)
Comment (streamer45) [reply service/ws/utils.go:20]: The reason is I am a firm believer of [A little copying is better than a little dependency.](https://go-proverbs.github.io/)
We don't want to couple the service with server's `model` if we can avoid it cause they don't share anything.

File: service/ws/server.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			s.log.Error("ws read failed", mlog.Err(err))
+			break
+		}
+
+		var msgType MessageType
+		if mt == websocket.TextMessage {
+			msgType = TextMessage
+		} else if mt == websocket.BinaryMessage {
+			msgType = BinaryMessage
+		}
```
[the remark this answers is not in the document]
Comment (streamer45) [reply service/ws/server.go:114]: We could, I just didn't know whether we had to move to a switch eventually so didn't really optimize :)

File: service/ws/server.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	for msg := range s.sendCh {
+		conn := s.getConn(msg.ConnID)
+		if conn == nil {
+			s.log.Error("failed to get conn for sending", mlog.String("connID", msg.ConnID))
+			continue
+		}
+
+		var msgType int
+		if msg.Type == TextMessage {
+			msgType = websocket.TextMessage
```
[the remark this answers is not in the document]
Comment (streamer45) [reply service/ws/server.go:145]: This is still valuable feedback especially as I am big fan of code conciseness.

**Comment by streamer45** [comment]
@cpoile I am going to merge this as I'll be addressing your feedback in the next PR.

**Pull request merged**